packages/react/src/App.ts

```diff
@@ -25,17 +26,25 @@ export default function App({
     )
   }, [])
 
+  useLayoutEffect(() => {
+    if (!current.preserveScroll) {
+      router.resetScrollPositions()
+    }
+  }, [current])
+
   useEffect(() => {
     router.init({
       initialPage,
       resolveComponent,
-      swapComponent: async ({ component, page, preserveState }) => {
+      swapComponent: async ({ component, page, preserveState, preserveScroll }) => {
         setCurrent((current) => ({
           component,
           page,
           key: preserveState ? current.key : Date.now(),
+          preserveScroll: !!preserveScroll,
         }))
       },
+      handleScrollResetExternally: true,
```

Comment:
Alternatively this property could take a callback that the router could call internally.